### ISSUE TITLE
src/dns.c: not checking return value suspected by coverity (new finding)

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#1126]
```
*** CID 1446655:  Error handling issues  (CHECKED_RETURN)
/src/dns.c: 1019 in dns_session_new()
1013     
1014            ds->tx_ring_area = pool_alloc(dns_msg_buf);
1015            if (!ds->tx_ring_area)
1016                    goto error;
1017     
1018            ring_init(&ds->ring, ds->tx_ring_area, DNS_TCP_MSG_RING_MAX_SIZE);
>>>     CID 1446655:  Error handling issues  (CHECKED_RETURN)
>>>     Calling "ring_attach" without checking return value (as is done elsewhere 4 out of 5 times).
1019            ring_attach(&ds->ring);
1020     
1021            if ((ds->task_exp = task_new(tid_bit)) == NULL)
1022                    goto error;
1023     
1024            ds->task_exp->process = dns_process_query_exp;
```

**Comment by EmericBr** [comment]
[0004-BUG-MINOR-dns-fix-ring-attach-control-on-dns_session.patch.txt](https://github.com/haproxy/haproxy/files/5982008/0004-BUG-MINOR-dns-fix-ring-attach-control-on-dns_session.patch.txt)
Adds a control on ring_attach even if it can not fail.

**Comment by git001** [comment]
How about to add some error information in the "should not happen" block to see when it happens even it should not happen :smile:

**Comment by TimWolla** [comment]
> How about to add some error information in the "should not happen" block to see when it happens even it should not happen smile

I generally agree with that. Maybe a `BUG_ON` should be placed there instead? It is clear to the static analyzers, because we run them with DEBUG_USE_ABORT and it will also fail loudly when the assumptions (“should”) are violated.

**Comment by wtarreau** [comment]
I totally agree. The problem with adding code to control things that "should not happen" is that the simple fact that code was written for this suggests this might happen, and that nobody really checked the impact. The comment does not help here. For me there are two solutions to deal with the "should not happen" case:
  - either a comment is written saying "cannot happen" and NO code is written for this
  - or it's a "should not" because we're uncertain and then a BUG_ON is the right way to check it.

Here Emeric says "it cannot fail", I'm fine with this, we have to trust the code's author more than the code analyser. But in this case we can also silence the analyser (either by marking it false positive or by consuming the return value).

**Comment by chipitsine** [comment]
we can suppress coverity with `// coverity ` syntax.
or explicitly (like I do now)

**Comment by wtarreau** [comment]
I'm not fond of having to deal with each and every external tool in the code. I'd rather please all of them at once with code that either does the right thing, or purposely hides what it's doing from their scrutiny.

**Comment by capflam** [comment]
Is there anything else to do on this issue ?

**Comment by chipitsine** [comment]
Coverity is happy, Willy is not

**Comment by wtarreau** [comment]
We can also use `DISGUISE` to stop the analysis by hiding the return, it was made for this and is already used at a few places where the compiler used to complain that the return as not checked. That would give `DISGUISE(ring_attach(&ds->ring));` with Emeric's comment about the fact that by definition it cannot fail.
Or maybe we can check if there is any single case where ring_attach() could fail and if it's justified to do it elsewhere, and if not, maybe even change it not to return anything.

**Comment by capflam** [comment]
From what I can see, `ring_attach()` may fail depending on the context. But not here because it is a freshly initialized ring. Thus there is no watcher on it. `ring_attach()` only fails it there are more than 255 watchers. A `DISGUISE()` is probably a good option.

**Comment by capflam** [comment]
Ok, fixed now with a`DISGUISE()` (1a1b674c2).
